File: 54-spiral-matrix/spiral-matrix.ts
function spiralOrder(matrix: number[][]): number[] {
    const r = matrix.length;
    const c = matrix[0].length;
    const n = r * c;
    let count = 0;
    const ans = [];
    let top = 0;
    let left = 0;
    let bottom = r - 1;
    let right = c - 1;
    while (count < n) {
        for (let i = left; i <= right; i++) {
            ans.push(matrix[top][i]);
            count++;
        }
        for (let i = top + 1; i <= bottom; i++) {
            ans.push(matrix[i][right]);
            count++;
        }
        if (top != bottom) {
            for (let i = right - 1; i >= left; i--) {
                ans.push(matrix[bottom][i]);
                count++;
            }
        }
        // console.log(ans);
        if (left != right) {
            for (let i = bottom - 1; i >= top + 1; i--) {
                ans.push(matrix[i][left]);
                count++;
            }
        }
        top ++;
        left ++;
        bottom --;
        right --;
    }
    return ans;
};

File: 54-spiral-matrix/spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        int rowStart = 0;
        int colStart = 0;
        int rowEnd = n - 1;
        int colEnd = m - 1;
        List<Integer> ans = new ArrayList<>();

        while (rowStart <= rowEnd && colStart <= colEnd) {
            for (int i = colStart; i <= colEnd; i++) {
                ans.add(matrix[rowStart][i]);
            }
            for (int i = rowStart + 1; i <= rowEnd; i++) {
                ans.add(matrix[i][colEnd]);
            }
            if (rowStart != rowEnd) {
                for (int i = colEnd - 1; i >= colStart; i--) {
                    ans.add(matrix[rowEnd][i]);
                }
            }
            if (colStart != colEnd) {
                for (int i = rowEnd - 1; i >= rowStart + 1; i--) {
                    ans.add(matrix[i][colStart]);
                }
            }
            rowStart++;
            colStart++;
            rowEnd--;
            colEnd--;
        }

        return ans;
    }
}